File: Beecrowd/Beginner/_3306_Search and Change/main.cpp
#include <iostream>
#include <vector>
#include <algorithm> 

using namespace std;

int main() {
    int tam_vetor, n_queries;
    cin >> tam_vetor >> n_queries;
    
    vector<int> vetor(tam_vetor);
    for (int i = 0; i < tam_vetor; i++) {
        cin >> vetor[i];
    }
    
    for (int i = 1; i <= n_queries; i++) {
        vector<int> query;
        int val;
        while (cin >> val) {
            query.push_back(val);
            if (cin.peek() == '\n' || cin.eof()) break;
        }
        
        if (query.size() < 3) {
            cout << i << endl;
        } else if (query[0] == 1) {
            for (int c = query[1] - 1; c < query[2]; c++) {
                vetor[c] += query[3];
            }
        } else {
            int result = vetor[query[1] - 1];
            for (int c = query[1]; c < query[2]; c++) {
                result = __gcd(result, vetor[c]);
            }
            cout << result << endl;
        }
    }
    return 0;
}

File: Beecrowd/Beginner/_3306_Search and Change/main.py
tam_vetor, n_queries = map(int, input().split()) 
vetor = list(map(int, input().split()))

def gcd(a: int, b: int) -> int:
    while b != 0:
        a, b = b, a % b
    return a

try:
    for _ in range(1,n_queries+1):
        querie = list(map(int, input().split()))
        if len(querie) < 3:
            print(_)
        elif querie[0] == 1:
            for c in range(querie[1]-1,querie[2]):
                vetor[c] += querie[3]
        else:
            result = vetor[querie[1]]
            for c in range(querie[1]-1,querie[2]):
                result = gcd(result,vetor[c])
            print(result)
except EOFError:
    print(n_queries)
    pass


   